src/Control/Effect/NonDet.hs

```diff
@@ -42,53 +43,19 @@ instance (Alternative f, Monad f, Traversable f, Carrier sig m, Effect sig, Appl
 --   prop> run (runNonDetOnce (asum (map pure (repeat a)))) == [a]
 --   prop> run (runNonDetOnce (asum (map pure (repeat a)))) == Just a
 runNonDetOnce :: (Alternative f, Monad f, Traversable f, Carrier sig m, Effect sig, Monad m) => Eff (OnceC f m) a -> m (f a)
-runNonDetOnce = runOnceC . interpret
+runNonDetOnce = runNonDet . runCull . cull . runOnceC . interpret
 
-newtype OnceC f m a = OnceC { runOnceC :: m (f a) }
+newtype OnceC f m a = OnceC { runOnceC :: Eff (CullC (Eff (AltC f m))) a }
 
-instance (Alternative f, Monad f, Traversable f, Carrier sig m, Effect sig, Monad m) => Carrier (NonDet :+: sig) (OnceC f m) where
-  ret a = OnceC (ret (pure a))
-  eff = OnceC . handleSum (eff . handleTraversable runOnceC) (\case
-    Empty    -> ret empty
-    Choose k -> do
-      l <- runOnceC (k True)
-      if null l then
-        runOnceC (k False)
-      else
-        pure l)
-
-
--- | The result of a nondeterministic branch of a computation.
---
---   'Branch' can be used to define 'NonDet' carriers which control nondeterminism in some specific way, e.g. pruning branches according to some specific heuristic.
-data Branch m e a
-  = None e
-  | Pure a
-  | Alt (m a) (m a)
-  deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
-
--- | Case analysis for 'Branch', taking a value to use for 'Cut', a value to use for 'None', and a function to apply to the contents of 'Pure'.
---
---   prop> branch None Pure Alt a == (a :: Branch e [] a)
---   prop> branch (applyFun f) (applyFun g) (applyFun2 h) (None a :: Branch [] a) == applyFun f a
---   prop> branch (applyFun f) (applyFun g) (applyFun2 h) (Pure a :: Branch [] a) == applyFun g a
---   prop> branch (applyFun f) (applyFun g) (applyFun2 h) (Alt a b :: Branch [] a) == applyFun2 h a b
-branch :: (e -> a) -> (b -> a) -> (m b -> m b -> a) -> Branch m e b -> a
-branch f _ _ (None a)  = f a
-branch _ f _ (Pure a)  = f a
-branch _ _ f (Alt a b) = f a b
-{-# INLINE branch #-}
-
--- | Interpret a 'Branch' into an underlying 'Alternative' context.
-runBranch :: Alternative m => (e -> m a) -> Branch m e a -> m a
-runBranch f = branch f pure (<|>)
-{-# INLINE runBranch #-}
+instance (Alternative f, Carrier sig m, Effect sig, Traversable f, Monad f, Monad m) => Carrier (NonDet :+: sig) (OnceC f m) where
+  ret = OnceC . ret
+  eff = OnceC . handleSum (eff . R . R . R . handleCoercible) (\case
+    Empty    -> empty
+    Choose k -> runOnceC (k True) <|> runOnceC (k False))
```

Comment:
We can just pass through to the underlying `NonDet` provided by `CullC`, and in turn provided by the underlying `AltC`.